server/schedule/region_splitter.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2020 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedule
+
+import (
+	"bytes"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/logutil"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule/opt"
+	"go.uber.org/zap"
+)
+
+// SplitRegionsHandler used to handle region splitting
+// TODO: support initialize splitRegionsHandler
+type SplitRegionsHandler interface {
+	SplitRegionByKeys(region *core.RegionInfo, splitKeys [][]byte) error
+	WatchRegionsByKeyRange(startKey, endKey []byte, timeout, watchInterval time.Duration) []uint64
+}
+
+// RegionSplitter handles split regions
+type RegionSplitter struct {
+	cluster opt.Cluster
+	handler SplitRegionsHandler
+}
+
+// NewRegionSplitter return a region splitter
+func NewRegionSplitter(cluster opt.Cluster, handler SplitRegionsHandler) *RegionSplitter {
+	return &RegionSplitter{
+		cluster: cluster,
+		handler: handler,
+	}
+}
+
+// SplitRegions support splitRegions by given split keys.
+func (r *RegionSplitter) SplitRegions(splitKeys [][]byte, retryLimit int) (int, []uint64) {
+	if len(splitKeys) < 1 {
+		return 0, nil
+	}
+	unprocessedKeys := splitKeys
+	newRegions := make(map[uint64]struct{}, len(splitKeys))
+	for i := 0; i <= retryLimit; i++ {
+		if i > 0 {
+			//TODO: sleep for a while
+			time.Sleep(500 * time.Millisecond)
+		}
+		unprocessedKeys = r.splitRegionsByKeys(unprocessedKeys, newRegions)
+		if len(unprocessedKeys) < 1 {
+			break
+		}
+	}
+	returned := make([]uint64, 0, len(newRegions))
+	for regionID := range newRegions {
+		returned = append(returned, regionID)
+	}
+	return 100 - len(unprocessedKeys)*100/len(splitKeys), returned
+}
+
+func (r *RegionSplitter) splitRegionsByKeys(splitKeys [][]byte, newRegions map[uint64]struct{}) [][]byte {
+	//TODO: support batch limit
+	groupKeys, unProcessedKeys := r.groupKeysByRegion(splitKeys)
+	for regionID, keys := range groupKeys {
+		region := r.cluster.GetRegion(regionID)
+		// TODO: assert region is not nil
+		// TODO: assert leader exists
+		// TODO: assert region replicated
+		// TODO: assert region not hot
```

Comment:
too many todos= =

Reply:
I use TODO here in order not to make this PR too big. The detailed implement could be referred here https://github.com/tikv/pd/pull/3072.

Reply:
ok